### PR TITLE
[vernac] [ast] Make location info an attribute of vernaculars.

### PR DESCRIPTION
**Pull request by ejgallego**
Whether locations where attached to vernacs has been a bit of a mess for a while; we try to be more uniform in the style of the rest of the AST.

### CODE REVIEW

File: vernac/ppvernac.ml [hunk 1]
```diff
@@ -1283,5 +1283,5 @@ let pr_vernac_attributes =
       return (keyword "Fail" ++ spc() ++ pr_vernac_control v)
 
     let pr_vernac v =
-      try pr_vernac_control v
+      try pr_vernac_control v.v
```
Comment (SkySkimmer) [inline vernac/ppvernac.ml]: Any reason why pr_vernac_control doesn't take the cast?

File: vernac/g_vernac.mlg [hunk 2]
```diff
@@ -148,7 +148,7 @@ GRAMMAR EXTEND Gram
   ;
 
   located_vernac:
-    [ [ v = vernac_control -> { CAst.make ~loc v } ] ]
+    [ [ v = vernac_control -> { v } ] ]
```
Comment (SkySkimmer) [inline vernac/g_vernac.mlg]: Any reason to keep `located_vernac`?

File: vernac/ppvernac.ml (hunk 1 above)
Comment (ejgallego) [reply vernac/ppvernac.ml]: Fixed thanks.

File: vernac/g_vernac.mlg (hunk 2 above)
Comment (ejgallego) [reply vernac/g_vernac.mlg]: Indeed, thanks.

File: ide/idetop.ml [hunk 3]
```diff
@@ -64,22 +64,22 @@ let is_known_option cmd = match Vernacprop.under_control cmd with
 
 (** Check whether a command is forbidden in the IDE *)
 
-let ide_cmd_checks ~last_valid {CAst.loc;v=ast} =
+let ide_cmd_checks ~last_valid ({CAst.loc;v=ast} as cmd) =
```
Comment (SkySkimmer) [inline ide/idetop.ml]: `ast` isn't used here

File: ide/idetop.ml [hunk 4]
```diff
@@ -64,22 +64,22 @@ let is_known_option cmd = match Vernacprop.under_control cmd with
...
       let (e, info) = CErrors.push e in
       let info = Stateid.add info ~valid:last_valid Stateid.dummy in
       Exninfo.raise ~info e
   in
-  if is_debug ast then
+  if is_debug cmd then
     user_error "Debug mode not available in the IDE"
 
-let ide_cmd_warns ~id {CAst.loc;v=ast} =
+let ide_cmd_warns ~id ({CAst.loc;v=ast} as cmd) =
```
Comment (SkySkimmer) [inline ide/idetop.ml]: `ast` isn't used here

File: stm/stm.ml [hunk 5]
```diff
@@ -2307,7 +2307,7 @@ let collect_proof keep cur hd brkind id =
         (try
           let name, hint = name ids, get_hint_ctx loc  in
           let t, v = proof_no_using last in
-          v.expr <- VernacExpr([], VernacProof(t, Some hint));
+          v.expr <- CAst.make @@ VernacExpr([], VernacProof(t, Some hint));
```
Comment (SkySkimmer) [inline stm/stm.ml]: Doesn't this cause the loc to be lost?

File: toplevel/vernac.ml [hunk 6]
```diff
@@ -20,12 +20,12 @@ open Vernacprop
...
    (the exceptions are explained only at the toplevel). *)
 
-let checknav_simple {CAst.loc;v=cmd} =
+let checknav_simple ({CAst.loc;v=_} as cmd) =
   if is_navigation_vernac cmd && not (is_reset cmd) then
     CErrors.user_err ?loc (str "Navigation commands forbidden in files.")
 
-let checknav_deep {CAst.loc;v=ast} =
-  if is_deep_navigation_vernac ast then
+let checknav_deep ({CAst.loc;v=ast} as cmd) =
```
Comment (SkySkimmer) [inline toplevel/vernac.ml]: `ast` isn't used here

**Review by SkySkimmer — COMMENTED** [review]
In a few places, `let foo = function Ba -> ...` is replaced by `let foo v = CAst.with_val (function Ba -> ...) v`.
It might be nicer to do `let foo v = v |> CAst.with_val (function Ba -> ...)` instead.

**Comment by ejgallego** [comment]
Merci for the corrections, fixed.

> In a few places, `let foo = function Ba -> ...` is replaced by `let foo v = CAst.with_val (function Ba -> ...) v`.
> It might be nicer to do `let foo v = v |> CAst.with_val (function Ba -> ...)` instead.

Why is it nicer? I see no difference honestly [as we are forced to eta-expand anyways due to the let-rec]

I can change it if you want, I don't mind a form or the other.

**Comment by SkySkimmer** [comment]
>Why is it nicer?

It's easier to see what the with_val is applied to since it's not multiple lines later IMO.

**Comment by ejgallego** [comment]
> It's easier to see what the with_val is applied to since it's not multiple lines later IMO.

Ok, I'll change it if you prefer; note that there are many more occurrences in the codebase of this pattern that I won't be changing.

**Comment by ejgallego** [comment]
This is IMO ready, needs assignee.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
Fairly consensual change. I'll merge this tomorrow if nobody complains.

**Comment by ejgallego** [comment]
Merci.

**Pull request merged**